rfd/0129-discovery-name-templating.md

```diff
@@ -0,0 +1,395 @@
+For example, base64 encode it, make the string lowercase, and replace the
+`[+/=]` characters with valid characters, maybe even truncating the result:
+(another heuristic, although less likely to collide names):
+
+```sh
+$ echo "weird-)(-group-name" | base64 | sed 's#[+/=]#x#g' | tr '[:upper:]' '[:lower:]' | cut -c1-8 
+d2vpcmqt
+$ echo "other-weird-)(-group-name" | base64 | sed 's#[+/=]#x#g' | tr '[:upper:]' '[:lower:]' | cut -c1-8 
+b3rozxit
+```
+
+- `foo-aks-d2vpcmqt-11111111-1111-1111-1111-111111111111`
+- `foo-aks-b3rozxit-11111111-1111-1111-1111-111111111111`
+- ...
+
+Each database name will be unique, since `foo` must be globally unique among
+all Azure MySQL databases and globally unique among all Azure Postgres databases.
+
+Even if a new database type is added that doesn't have this globally unique
+name property, the resource group name and subscription ID will avoid name
+collisions, and the databases will be distinguished from databases in other
+clouds.
+
+Likewise, the discovered AKS clusters will avoid colliding with other kube
+clusters in Azure or other clouds.
+
+This naming convention does not violate our database name validation regex,
+`^[a-z]([-a-z0-9]*[a-z0-9])?$`,
+and does not violate our kube cluster name validation regex `^[a-zA-Z0-9._-]+$`.
+
+#### GCP Discovery
+
+GCP discovery currently supports discovering only GKE kube clusters.
+
+GKE cluster names are unique within the same GCP project ID and location/zone.
+
+The discovery naming convention for GKE clusters shall be to append a suffix to
+the cluster name that includes:
+
+- Name of the Teleport GCP matcher type
+  - `gke`
+- GCP project ID
+  - These can be custom, but will only consist of characters, digits, hyphens.
+- GCP location
+
+```yaml
+    gcp:
+    - types: ["gke"]
+      locations: ["us-west1", "us-west2"]
+      tags:
+        "*": "*"
+      project_ids: ["my-project"]
+```
+
+If the discovery service is configured like the above, the discovery agent will
+discover GCP GKE kube clusters in "my-project" in the `us-west1` and `us-west2`
+locations.
+
+Now suppose GKE clusters named `foo` exist in each region.
+If the discovery service applies the new naming convention, the discovered
+resources should be named:
+
+- `foo-gke-us-west1-my-project`
+- `foo-gke-us-west2-my-project`
+
+This naming convention avoids name collisions between GKE clusters and does not
+collide with discovered AWS/Azure clusters.
+
+This naming convention does not violate our kube cluster name validation regex:
+`^[a-zA-Z0-9._-]+$`
+
+### `tsh` UX
+
+Users will be frustrated if they are forced to type out verbose resource names
+when using `tsh`.
+To avoid this poor UX, sub-commands should support prefix resource names or label
+matching to identify resources.
+
+The same UX should apply to all `tsh` sub-commands that take a resource name
+argument. These commands shall support
+`tsh <sub-command> [name | prefix] [key1=value1,key2=value2,...]` syntax:
+
+- `tsh db login`
```

Comment:
Can we include `tsh app xxx` commands in the list?

Reply:
For UX consistency, yes. I only excluded it from the list because we don't have app auto-discovery

Reply:
2a3bd6e456f7e5119e9f32273d3921b5a4d27832